core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1029,43 +1047,32 @@ private void sendRequestToChannel(DiscoveryNode node, final Channel targetChanne
             stream.writeString(action);
             BytesReference message = buildMessage(requestId, status, node.getVersion(), request, stream, bStream);
             final TransportRequestOptions finalOptions = options;
-            Runnable onRequestSent = () -> { // this might be called in a different thread
-                try {
-                    toRelease.close();
-                } finally {
-                    transportServiceAdapter.onRequestSent(node, requestId, action, request, finalOptions);
-                }
-            };
-            addedReleaseListener = internalSendMessage(targetChannel, message, onRequestSent);
+            // this might be called in a different thread
+            SendListener onRequestSent = new SendListener(toRelease,
+                () -> transportServiceAdapter.onRequestSent(node, requestId, action, request, finalOptions));
+            internalSendMessage(targetChannel, message, onRequestSent);
         } finally {
             IOUtils.close(stream);
-            if (!addedReleaseListener) {
-                toRelease.close();
-            }
         }
     }
 
     /**
-     * sends a message view the given channel, using the given callbacks.
-     *
-     * @return true if the message was successfully sent or false when an error occurred and the error hanlding logic was activated
-     *
+     * sends a message to the given channel, using the given callbacks.
      */
-    private boolean internalSendMessage(Channel targetChannel, BytesReference message, Runnable onRequestSent) throws IOException {
-        boolean success;
+    private void internalSendMessage(Channel targetChannel, BytesReference message, ActionListener<Channel> listener) {
```

Comment:
maybe we can optimize this a little, what about this: Instead of having a static method on `ActionListener` we add a base class `NotifyOnceListener` that has 2 `innerOnFailure` and `innerOnResponse` abstract methods. 
That way we require `NotifyOnceListener` on this method and subclass it from `SendListener`. That way we have it guaranteed that all listeners coming in can be called twice and we don't need to to any extra wrapping hence not additional object creation?!